### ISSUE TITLE
Box回归损失为负数

### ISSUE THREAD

**Issue by HITyu4** [fppccc/Powerful-IoU#5]
iou = 1-piou(pred_bboxes[fg_mask], target_bboxes[fg_mask], xywh=False, PIoU2=True).clamp_(0)
box_loss最终计算的结果为负数，是否正常

**Comment by fppccc** [comment]
您好！感谢您对本工作的关注。这里已经使用了clamp_()函数进行截断，box_loss的计算结果不应该再出现负数了。如果可以的话，请您将后一行的代码也发给我看一下。

**Comment by HITyu4** [comment]
# pxy, pwh, _, pcls = pi[b, a, gj, gi].tensor_split((2, 4, 5), dim=1)  # faster, requires torch 1.8.0
                pxy, pwh, _, pcls = pi[b, a, gj, gi].split((2, 2, 1, self.nc), 1)  # target-subset of predictions

                # Regression
                pxy = pxy.sigmoid() * 2 - 0.5
                pwh = (pwh.sigmoid() * 2) ** 2 * anchors[i]
                pbox = torch.cat((pxy, pwh), 1)  # predicted box
                # iou = bbox_iou(pbox, tbox[i], FIoU=True).squeeze()  # iou(prediction, target)原始
                iou =  1-piou(pbox, tbox[i], xywh=False, PIoU2=True).clamp_(0)
                lbox += (1.0 - iou).mean()  # iou loss

                # Objectness
                iou = iou.detach().clamp(0).type(tobj.dtype) 您好改好了这个bug，想确认一下这个损失这么用对吗，效果好像不是很理想

**Comment by fppccc** [comment]
clamp(0)应改为clamp_(0)。
因为Ultralytics的版本迭代，您发的代码与我使用的版本略有不同，如果需要进一步讨论的话，可以将loss.py文件发送至canliu@whu.edu.cn
此外，我个人认为适用于每个检测器和数据集的最佳损失函数都是不同的，您可以通过实验尝试来选择最适合您的实验的损失函数。
